File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>paths;
    vector<int>t;
    void check(TreeNode* root, int targetSum, vector<int> t){
        if(root==NULL) return ;
        if(!root->left and !root->right){
            if(root->val==targetSum){
                t.push_back(root->val);
                paths.push_back(t);
            }
        }
        t.push_back(root->val);
        check(root->left,targetSum-root->val,t);
        check(root->right,targetSum-root->val,t);
        return ;
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        check(root, targetSum, t);
        return paths;
    }
};

File: 113-path-sum-ii/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>>f=new ArrayList<>();
    public void check(TreeNode r, int targetsum, List<Integer>t){
        if(r==null) return ;
        t.add(r.val);
        if(targetsum==r.val && r.left==null && r.right==null){
            if(!t.isEmpty())
                f.add(new ArrayList<>(t));
        }
        check(r.left,targetsum-r.val,t);
        check(r.right,targetsum-r.val,t);
        t.remove(t.size()-1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer>t=new ArrayList<>();
        check(root,targetSum,t);
        return f;
    }
}